### ISSUE TITLE
Frame rate stuttering on macOS

### ISSUE THREAD

**Issue by fredizzimo** [neovide/neovide#2093]
On some macOS systems the frame rate stutters a bit due to a buggy vsync implementation. If you have that issu, it's recommended to:
1. Start Neovide with the command `--no-vsync` to disable the vsync and switch to a software based timer. 
2. Set `g:neovide_refresh_rate` https://neovide.dev/configuration.html#refresh-rate to the same as your current monitor refresh rate. This combination should make it reasonably smooth, but if not, you might want to try different refresh rates to see what works for you.

This will be properly fixed by:
* https://github.com/rust-windowing/winit/issues/2412

But currently the winit team don't have the manpower to implement it, so contributors are welcome. @crupest is also working on a quick fix for Neovide itself, at least that's the status at the time of writing, see
* https://github.com/neovide/neovide/issues/2073, so any additional effort should be discussed with them first.

The vsync issues are well known and discussed here:
* https://github.com/libsdl-org/SDL/issues/4918
* https://github.com/glfw/glfw/issues/2249
* https://github.com/libsdl-org/SDL/issues/4471
* https://github.com/libsdl-org/SDL/issues/1871
* https://github.com/glfw/glfw/issues/2249

And also on or Discord server: 
* https://discord.com/channels/896132176091955261/896263985614168125/1132361827095953408
* https://discord.com/channels/896132176091955261/896263985614168125/1142096375178342480
* https://discord.com/channels/896132176091955261/896263894086066216/1152521461152628736

**Comment by crupest** [comment]
Hi! Sorry for late reply. I'm busy these two weeks. But luckily, I can spend this weekend on this problem. If you happen to be a macOS developer and want to help, here are some points:

1. I will try to fix neovide without touch winit as much as possible. For the screen change problem, I think I can workaround by move event.

2. https://github.com/rust-windowing/winit/issues/2640 It is a little wired this issue is fixed on iOS but not macOS. I'm not sure whether this is related to neovide, but I think it is important to fix this issue. However, it needs a lot change on winit codebase.

3. As @fredizzimo says, the best is to implement https://github.com/rust-windowing/winit/issues/2412

I'll work on 1 specifically. But you can really help on 2 and 3.

Best regards!

**Comment by fredizzimo** [comment]
@crupest. Thank you. Yes, I agree that fixing the issue for just Neovide seems like the best approach to start with.

2. Should not affect us at the moment, since the update loop switches to polling when animating https://github.com/neovide/neovide/blob/5743356d654dbeaf213c0da3e2988c4d89c218e1/src/window/update_loop.rs#L165. But it probably need to be fixed in combination with the vsync source.

**Comment by crupest** [comment]
@fredizzimo Hey! Would you mind have a look on this. The main branch can't run on macos now. Did I miss something?

```
[flexi_logger][ERRCODE::Time] flexi_logger has to work with UTC rather than with local time, caused by IndeterminateOffset
    See https://docs.rs/flexi_logger/latest/flexi_logger/error_info/index.html#time
Neovide panicked with the message 'invalid message send to -[NSOpenGLContext CGLContextObj]: expected return to have type code '^{_CGLContextObject=^{__GLIContextRec}{__GLIFunctionDispatchRec=^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?}^{_CGLPrivateObject}^v}', but found '^{_CGLContextObject=}''. (File: /Users/crupest/.cargo/registry/src/index.crates.io-6f17d22bba15001f/glutin-0.31.0/src/api/cgl/appkit.rs; Line: 38, Column: 1)
This is a bug and we would love for it to be reported to https://github.com/neovide/neovide/issues
   0: backtrace::backtrace::libunwind::trace
             at /Users/crupest/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.69/src/backtrace/libunwind.rs:93:5
      backtrace::backtrace::trace_unsynchronized
             at /Users/crupest/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.69/src/backtrace/mod.rs:66:5
   1: backtrace::backtrace::trace
             at /Users/crupest/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.69/src/backtrace/mod.rs:53:14
   2: backtrace::capture::Backtrace::create
             at /Users/crupest/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.69/src/capture.rs:176:9
   3: backtrace::capture::Backtrace::new
             at /Users/crupest/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.69/src/capture.rs:140:22
   4: neovide::main::{{closure}}
             at src/main.rs:75:25
   5: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/alloc/src/boxed.rs:2021:9
      std::panicking::rust_panic_with_hook
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:711:13
   6: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:599:13
   7: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/sys_common/backtrace.rs:170:18
   8: rust_begin_unwind
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:595:5
   9: core::panicking::panic_fmt
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/panicking.rs:67:14
  10: objc2::message::panic_verify
             at /Users/crupest/.cargo/registry/src/index.crates.io-6f17d22bba15001f/objc2-0.4.1/src/message/mod.rs:83:5
  11: objc2::message::msg_send_check
             at /Users/crupest/.cargo/registry/src/index.crates.io-6f17d22bba15001f/objc2-0.4.1/src/message/mod.rs:71:5
  12: objc2::message::MessageReceiver::send_message
             at /Users/crupest/.cargo/registry/src/index.crates.io-6f17d22bba15001f/objc2-0.4.1/src/message/mod.rs:231:13
  13: glutin::api::cgl::appkit::NSOpenGLContext::CGLContextObj
             at /Users/crupest/.cargo/registry/src/index.crates.io-6f17d22bba15001f/objc2-0.4.1/src/macros/__attribute_helpers.rs:126:21
  14: glutin::api::cgl::context::<impl glutin::api::cgl::display::Display>::create_context
             at /Users/crupest/.cargo/registry/src/index.crates.io-6f17d22bba15001f/glutin-0.31.0/src/api/cgl/context.rs:57:33
  15: <glutin::display::Display as glutin::display::GlDisplay>::create_context
             at /Users/crupest/.cargo/registry/src/index.crates.io-6f17d22bba15001f/glutin-0.31.0/src/display.rs:311:43
  16: neovide::renderer::opengl::build_context
             at src/renderer/opengl.rs:118:28
  17: neovide::window::window_wrapper::WinitWindowWrapper::new
             at src/window/window_wrapper.rs:77:32
  18: neovide::window::main_loop
             at src/window/mod.rs:312:30
  19: neovide::main
             at src/main.rs:95:13
  20: core::ops::function::FnOnce::call_once
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/ops/function.rs:250:5
  21: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/sys_common/backtrace.rs:154:18
  22: std::rt::lang_start::{{closure}}
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/rt.rs:166:18
  23: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/ops/function.rs:284:13
      std::panicking::try::do_call
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:502:40
      std::panicking::try
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:466:19
      std::panic::catch_unwind
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panic.rs:142:14
      std::rt::lang_start_internal::{{closure}}
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/rt.rs:148:48
      std::panicking::try::do_call
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:502:40
      std::panicking::try
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:466:19
      std::panic::catch_unwind
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panic.rs:142:14
      std::rt::lang_start_internal
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/rt.rs:148:20
  24: std::rt::lang_start
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/rt.rs:165:17
  25: _main
  26: <unknown>
```

**Comment by crupest** [comment]
And in terminal

```
[flexi_logger][ERRCODE::Time] flexi_logger has to work with UTC rather than with local time, caused by IndeterminateOffset
    See https://docs.rs/flexi_logger/latest/flexi_logger/error_info/index.html#time
Neovide panicked with the message 'invalid message send to -[NSOpenGLContext CGLContextObj]: expected return to have type code '^{_CGLContextObject=^{__GLIContextRec}{__GLIFunctionDispatchRec=^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?^?}^{_CGLPrivateObject}^v}', but found '^{_CGLContextObject=}''. (File: /Users/crupest/.cargo/registry/src/index.crates.io-6f17d22bba15001f/glutin-0.31.0/src/api/cgl/appkit.rs; Line: 38, Column: 1)
```

**Comment by fredizzimo** [comment]
That was also reported in this issue, https://github.com/neovide/neovide/issues/2097, but I don't think it was the main cause of the issue itself.

It looks like `glutin`needs to be updated to this release https://github.com/rust-windowing/glutin/releases/tag/v0.31.1
* https://github.com/rust-windowing/glutin/issues/1640

**Comment by fredizzimo** [comment]
@crupest has now implemented vsync frame callback on macOS, so the frame stuttering issues on that should now be fixed. 

If you have previously used special command-line parameters and custom `g:neovide_refresh_rate` settings, it's recommended that you remove them and let everything be left on the defaults. The refresh rate should then be the same as the one you have configured for your monitor.

It would be nice if as many people as possible could test it out, and report any issues here.

**Comment by 9mm** [comment]
So when I set it explicitly, I get up to 2000 fps in the profiler, and i dont know the refresh rate of my monitor, let me check.

It says "ProMotion" which appears to be 120Hz

Now, if I do:

/neovide --no-idle --vsync

And open profiler, it hovers around 120fps with a "max" of about 9ms

If I do no-vsync and set fps very high in the config, it can go as high as 2000

So this appears to work, but my question is...

is having it RIGHT at 120 the "most" smooth it can be? if its right at the cusp, wouldnt it be occassionally skipping frames during movement etc? Is it better to say... double it, or even leave it at 1000 or something, so that quite literally every frame it gets an update, or am I thinking about this the wrong way?

Thanks, and great work!

**Comment by crupest** [comment]
I have no ideas about whether vsync or not is better. There is an option of vsync in most games but I never care about it and let it as default always. I only know what vsync is but not know much about what influence it will finally make. I just implement vsync following the guide of @fredizzimo . For me, there is no significant difference between these two options. And actually I was intended to fix another bug, #2073, at first, which finally leads me here.

**Comment by fredizzimo** [comment]
@9mm. If the vsync works properly it should be as smooth as it can be. 

There should be no skipped frames, unless it's caused by the computer being slow, since the frame callback implementation is supposed to tell Neovide exactly when to render so that no frames are missed. I'm not sure how the callback is implemented, but logically it should be called just after the previous frame is ready, so that Neovide has the maximum amount of time to prepare the next frame.

A max of 9ms is slightly too high, since 120 FPS is 8.33 ms, but that is probably compensated by the next frame being slightly too short. This kind of pattern is usually ok, and won't cause a missed frame due to the double and triple buffering the compositor implements. 

When you render at more than the vsync interval, the compositor will most likely use the latest frame available. This can actually cause more stuttering, unless the fps you render is exactly divisible by the vsync interval. For example if your vsync is 120 and you render at 250, then you will see every other frame, but every 12th frame it will skip over 2 frames.

In games, however you typically enable tearing when you render with a higher FPS, this way you get less latency, but the top of the screen and the bottom of the screen can display different frames, which can be seen as slight visual artefacts. But if the refresh rate is high enough it's usually not very noticeable, so many especially competitive gamers prefers it this way. It can also sometimes be useful to not sync to the vsync if your computer is too slow to render at the full frame rate, so rather than dropping the fame rate in half, which would happen with vsync, you can render at 110 FPS for example with a bit of tearing.

All that mostly applies to full screen applications, like games though, as mentioned I doubt the macOS compositor allows it for windowed mode applications like Neovide.

If you do feel like you get a smoother frame rate with for example 1000 FPS, then it might be worth profiling it to check what's going on. But I'm not sure we can do anything more about it, @crupest's implementation seem to be as good as it can get. But the most important is that it selects something that is close to completely smooth by default, without the need to tweak anything, people that want complete smoothness can try to tweak it themselves. It also uses less energy to render with a lower FPS, which is important for laptop users.

**Comment by 9mm** [comment]
Hmm interestingly i already notice its significantly choppy, somehow i didnt notice it before but now that im navigating through telesocpe windows.

default 120 FPS with vsync=1 is very choppy, much worse than the current homebrew neovim

**Comment by 9mm** [comment]
I can try to do a more official test later, that was just a quick test because i was surprised to see it lagging quite hard

**Comment by 9mm** [comment]
I also tried to make a loom but the FPS is too bad to notice

**Comment by fredizzimo** [comment]
I started to research if we are still doing something wrong and I found this https://developer.apple.com/videos/play/wwdc2021/10147/. The adaptive vsync is not relevant, but the second part is.

But I don’t see anything wrong, the callback is indeed giving us the maximum time to render. The only thing we could do better is to use the target timestamp instead of the running average we currently use for the timing, but I don’t think it would improve things very much, unless there’s constant thermal throttling or something.

But maybe a tracy profiling log could reveal something? I will give you instructions tomortow, @9mm, after I have updated Neovide to work with the current new version of Tracy.

**Comment by fredizzimo** [comment]
@9mm, https://github.com/neovide/neovide/pull/2149 adds support for the latest tracy profiler, which you can install through homebrew. 

You can then run neovide with `tracy-capture -o vsync.tracy & cargo run --release --features profiling -- --no-fork` to create a tracy capture ('vsync.tracy`) that you can attach here after you have reproduced the stuttering. The file should be quite small, even if you run for a long time, so you don't necessarily need to be able to reproduce it immediately. But it's of course easier for me to find thing in a shorter capture.

If you need several different captures, you can just replace `vsync.tracy` with another name, but keep the `tracy` extension, since it's easier to open it that way.

If you want to check the catpture yourself you can use the gui application, with instructions here https://github.com/wolfpld/tracy

**Comment by fredizzimo** [comment]
Hm. I might be wrong, something else seems to be going on

![image](https://github.com/neovide/neovide/assets/9946255/6c811657-3fc5-4797-95aa-6255d891be99)


Notice two very short frames (which are not actually drawn) between the ones that are drawn. I don't know why those are not being drawn.

Also re-opening, until we have figured this out.

**Comment by fredizzimo** [comment]
Upon further inspection, that is correct if you have set `neovide_cursor_animation_length` to 0 and typed during the capture. Is that possible @9mm? 

But on the other hand, if that's the case, there should be nothing that can stutter, since there's no animation.

**Comment by 9mm** [comment]
@fredizzimo yes thats correct:

```lua
  -- padding
  vim.g.neovide_padding_top    = 20
  vim.g.neovide_padding_bottom = 20
  vim.g.neovide_padding_right  = 20
  vim.g.neovide_padding_left   = 20

  -- display
  -- vim.g.neovide_transparency = 0.8
  -- vim.g.transparency = 0.8
  -- vim.g.neovide_background_color = "#0f0f12" .. alpha()
  -- vim.g.neovide_refresh_rate = 960 -- fps peaks over 1000+ (requires NEOVIDE_VSYNC=0)
  vim.g.neovide_refresh_rate_idle = 60
  vim.g.neovide_hide_mouse_when_typing = true
  vim.g.neovide_underline_stroke_scale = 1.2

  -- window
  vim.g.neovide_remember_window_size = true
  vim.g.neovide_confirm_quit = true

  -- animation
  vim.g.neovide_cursor_animation_length = 0
  vim.g.neovide_cursor_trail_size = 0
  vim.g.neovide_scroll_animation_length = 0
  vim.g.neovide_cursor_animate_command_line = false

  -- other
  -- for when it gets merged into main...
  -- vim.g.neovide_input_macos_option_key_is_meta = "OnlyLeft"
  vim.g.neovide_input_macos_alt_is_meta = true
```

**Comment by 9mm** [comment]
@fredizzimo specifically what is stuttering is if I open telescope with lots of files, and just hold down the "j" key to scroll as fast as possible, its extremely janky, where as when its maxed at 9999 fps its so smooth its beautiful

**Comment by fredizzimo** [comment]
Towards the end of the capture there appears to be some scrolling, but there's still no animation, but that's according to your configuration so that's probably correct.

Could you try with a `neovide_scroll_animation_length`  to 0.130? That should still be very fast but smooth. And a lot easier to diagnose, even if you don't end up liking it.

Maybe it would also be worth using `--no-idle` for this profiling.

**Comment by 9mm** [comment]
[vsync.tracy.zip](https://github.com/neovide/neovide/files/13476785/vsync.tracy.zip)

I definitely see the animation effect now, it does seem to be smoother by a lot, but now it has the "bouncy" effect which ultimately was why i turned it off (albeit very cool, sometimes distracting if im thinking)

**Comment by fredizzimo** [comment]
Yes, based on the tracy log, that seems to work correctly, with very minor jittering, but probably rendering exactly at the vsync rate in real life.

I think we can make it smooth even without animating, if we keep the render queue busy always rendering for maybe 100 ms after an input, until going idle. A little bit like an animation that does nothing. That way every frame will be rendered without drops when you hold down a key, since 100ms is longer than most key repeat rates.

I will prepare a new branch with those changes.

I'm not sure if it needs to be configurable or not, since someone might want to turn off all kinds of animation in order to preserve the battery life of a laptop for example. But I guess we can leave those micro-optimisations for later, the fixes I did in that PR is probably enough to compensate for the additional power requirements this would add. There are also more worthwhile optimisations, like only re-drawing what's changed, which would help more.

**Comment by crupest** [comment]
I may not have time to dig into what problems you have run into. But if you need, I'm willing to help, especially for implementing codes. Just @ me if you need.

**Comment by fredizzimo** [comment]
@crupest, I think I determined in https://github.com/neovide/neovide/pull/2151 that the swap and vsync wait need to be swapped around for improved latency. Now the swap happens after the vsync, which introduces one extra frame of latency, in practice there's almost two frames of latency from reading the inputs to the time the compositor gets it, and then there's usually one or more frame latency in the compositor still.

I need to think a bit more about that, but I think it actually always applies on all platforms. If `swap_buffer` actually blocks by itself, then our implementations should return immediately, so there should be no additional delays when waiting after that. 

I will do some experimenting and profiling to check if that's the case. I think it only ended up being this way because I had trouble getting the Windows implementation to never drop any frames and switched things around, but I never quite got it to completely work, so I gave up and left it in a state where it was reasonable smooth, even if switching them around would have been the same with better latency.
